File: 12-2019/Program.py
class Cidade:
def __init__(self, nome, estado):
self.nome = nome
self.estado = estado

def __str__(self):
return self.nome+ "-"+self.estado

def __eq__(self, other):
"""Overrides the default implementation"""
if isinstance(other, Cidade):
return self.nome == other.nome and self.estado == other.estado
return False







class Pessoa:
def __init__(self, nome, cidade):
self.nome = nome
self.cidade = cidade
self.seguidores=[]
self.seguindo=[]

def seguir(self, pessoa):
self.seguindo.append(pessoa)
pessoa.seguidores.append(self)

def seguidoresEmComum(self, pessoa):
return [seguidor for seguidor in self.seguidores if seguidor in pessoa.seguidores]

def seguidoresDe(self, cidade):
return sorted([seguidor for seguidor in self.seguidores if seguidor.cidade == cidade], key = lambda p : p.nome)

def __str__(self):
return self.nome + " de " + str(self.cidade)








def printSeguindo(pessoa):
print(f"{pessoa} segue {', '.join(str(x) for x in pessoa.seguindo) }");
def printSeguidoresEmComum(pessoa1,pessoa2):
print(f"{pessoa1} e {pessoa2} ambos são seguidos por: {', '.join(str(x) for x in pessoa1.seguidoresEmComum(pessoa2)) }");
def printSeguidoresDe(pessoa,cidade):
print(f"{pessoa} tem os seguintes seguidores de {cidade}: {', '.join(str(x) for x in pessoa.seguidoresDe(cidade)) }");

curitiba = Cidade("Curitiba","PR")
maringa = Cidade("Maringá","PR")
rio = Cidade("Rio de Janeiro","RJ")







tiago=Pessoa("Tiago",curitiba)
marcio=Pessoa("Marcio",curitiba)
wesley=Pessoa("Wesley",maringa)
jonas=Pessoa("Jonas",rio)





tiago.seguir(wesley)
tiago.seguir(marcio)
tiago.seguir(jonas)







jonas.seguir(wesley)
jonas.seguir(marcio)
jonas.seguir(tiago)







wesley.seguir(wesley)







marcio.seguir(jonas)








printSeguindo(tiago)
printSeguindo(marcio)
printSeguindo(jonas)
printSeguindo(wesley)







printSeguidoresEmComum(wesley,marcio)







printSeguidoresDe(jonas,curitiba)

File: 12-2019/Program.java
import java.util.List;
import java.util.Arrays;
import java.util.ArrayList;
import java.util.Comparator;
import java.util.stream.Collectors;

public class Program {
    public static void main(String args[]) {

        City curitiba = new City("Curitiba", "PR");
        City maringa = new City("Maringá", "PR");
        City rio = new City("Rio de Janeiro", "RJ");

        Person marcio = new Person("Marcio", curitiba);
        Person tiago = new Person("Tiago", curitiba);
        Person wesley = new Person("Wesley", maringa);
        Person jonas = new Person("Jonas", rio);

        tiago.follow(marcio);
        tiago.follow(wesley);
        tiago.follow(jonas);
        jonas.follow(marcio);
        jonas.follow(tiago);
        jonas.follow(wesley);
        marcio.follow(wesley);
        wesley.follow(jonas);

        System.out.println("2.1) Mostre no console todas as pessoas que cada pessoa está seguindo.\n");
        printPerson(tiago);
        printPerson(marcio);
        printPerson(wesley);
        printPerson(jonas);

        System.out.println("\n----------------------------------------------------------------------");
        System.out.println("3.1) Mostre no console os seguidores em comum entre Wesley e Marcio.\n");
        printCommon(marcio, wesley);

        System.out.println("\n----------------------------------------------------------------------");
        System.out.println("4.1) Mostre no Console os seguidores de Curitiba do Wesley.\n");
        printFollowersFrom(wesley, curitiba);

    }

    private static void printPerson(Person person) {
        System.out.println(person.getName() + " segue: " + Arrays.toString(person.getFollowing().toArray()));
    }

    private static void printCommon(Person personOne, Person personTwo) {
        System.out.println(personOne.getName() + " e " + personTwo.getName() + " são seguidos por: "
                + Arrays.toString(personOne.getCommonFollowers(personTwo).toArray()));
    }

    private static void printFollowersFrom(Person person, City city) {
        System.out.println(person.getName() + " é seguido pelas seguintes pessoas de " + city.getName() + ": "
                + Arrays.toString(person.getFollowersFrom(city).toArray()));
    }

}

class City {
    String name;
    String state;

    public City(String city, String state) {
        this.name = city;
        this.state = state;
    }

    public String getName() {
        return this.name;
    }

    public String getState() {
        return this.state;
    }

    @Override
    public String toString() {
        return this.name + "-" + this.state;
    }
}

class Person {
    String name;
    City city;
    List<Person> followers;
    List<Person> following;

    public Person(String name, City city) {
        this.name = name;
        this.city = city;
        followers = new ArrayList<Person>();
        following = new ArrayList<Person>();
    }

    public void follow(Person person) {
        this.following.add(person);
        person.followers.add(this);
    }

    public String getName() {
        return this.name;
    }

    public City getCity() {
        return this.city;
    }

    public List<Person> getFollowers() {
        return this.followers;
    }

    public List<Person> getFollowing() {
        return this.following;
    }

    public List<Person> getCommonFollowers(Person person) {
        List<Person> commonFollowers = new ArrayList<Person>();

        for (Person myFollower : this.followers) {
            for (Person otherFollower : person.getFollowers()) {
                if (myFollower.equals(otherFollower)) {
                    commonFollowers.add(myFollower);
                }
            }
        }
        return commonFollowers;
    }

    public List<Person> getFollowersFrom(City city) {
        return this.followers.stream()
        .filter(follower -> city.equals(follower.getCity()))
        .sorted(Comparator.comparing(Person::getName))
        .collect(Collectors.toList());
    }

    @Override
    public String toString() {
        return this.name + " de " + this.city;
    }

}